Sources/SpeziHealthKit/Sample Types/SampleType.swift

```diff
@@ -0,0 +1,199 @@
+    associatedtype _SampleType: HKSampleType
+}
+
+
+extension HKQuantitySample: _HKSampleWithSampleType {
+    public typealias _SampleType = HKQuantityType
+}
+
+extension HKCorrelation: _HKSampleWithSampleType {
+    public typealias _SampleType = HKCorrelationType
+}
+
+extension HKCategorySample: _HKSampleWithSampleType {
+    public typealias _SampleType = HKCategoryType
+}
+
+extension HKElectrocardiogram: _HKSampleWithSampleType {
+    public typealias _SampleType = HKElectrocardiogramType
+}
+
+extension HKAudiogramSample: _HKSampleWithSampleType {
+    public typealias _SampleType = HKAudiogramSampleType
+}
+
+
+
+// TODO is this actually needed?
+public protocol AnyHealthKitSampleType {
+    associatedtype Sample: _HKSampleWithSampleType
+    var hkSampleType: Sample._SampleType { get }
+}
+
+
+
+/// A sample type that can be used to query data from HealthKit.
+///
+/// This struct is a strongly-typed wrapper around `HKSampleType` that associates a sample type with its corresponding `HKSample` subclass.
+/// Additionally, it also provides some useful information for working with a sample type, such as e.g. the sample type's recommended display title
+public struct HealthKitSampleType<Sample: _HKSampleWithSampleType>: AnyHealthKitSampleType, Hashable, Identifiable, Sendable {
+    private enum Variant: Sendable {
+        /// - parameter displayUnit: The unit that should be used when displaying a sample of this type to the user
+        /// - parameter expectedValuesRange: The expected range of values we expect to see for this sample type, if applicable.
+        ///     The main purpose of this is to be able to e.g. adjust chart value ranges based on the specific sample types being visualised.
+        case quantity(displayUnit: HKUnit, expectedValuesRange: ClosedRange<Double>?) // TODO remove the range thing?!
+        /// - parameter displayUnit: The unit that should be used when displaying a sample belonging to a correlation of this type to the user.
+        ///    Depending on the specific correlation type, this value might be `nil`. (E.g., if the samples associated with the correlation don't all use the same unit.)
+        case correlation(displayUnit: HKUnit?)
+        case category
+        case electrocardiogram
+        case audiogram
+    }
+    
+    /// The underlying `HKSampleType`
+    public let hkSampleType: Sample._SampleType
+    /// The recommended user-displayable name of this sample type.
+    public let displayTitle: String
+    /// Variant-specific additional information.
+    private let variant: Variant
+    
+    public var id: String {
+        hkSampleType.identifier
+    }
+    
+    /// Creates a new ``HealthKitSampleType``.
+    /// - Note: Don't use this initializer directly. Instead, use one of the static factory methods: `HealthKitSampleType.quantity`, `.correlation`, and `.category`.
+    private init(_ hkSampleType: Sample._SampleType, displayTitle: String, variant: Variant) {
```

Comment:
Wondering if the the display title should rather be a `LocalizedStringResource` or similar localizable elements to enable this in the future or at least provide the option to use the displayTitles in the UI?

Reply:
yes, definitely.

there's multiple places where the whole `HealthKitSampleType` should get localized:
- anything that is user-displayed text (e.g. the displayTitles of the sample types)
- anything that displays a value in a unit (e.g., need to differentiate between US, metric, and the UK)

i'm currently already doing this a little bit in the `displayUnit` of the `height` sample type (though that is probably still a little wrong, since you can only express feet as a `HKUnit`, but not feet and inches).
once we add a couple more sample types, we'd also need to localize temperatures between C/F/etc.